### PR TITLE
[JBPM-9405] Provide API method to signal process instance with correlation key

### PR DESCRIPTION
**Pull request by elguardian**
Jira JBPM-9405

Added rest endpoints and interfaces.

depends on :
https://github.com/kiegroup/jbpm/pull/1767

### CODE REVIEW

**Comment by elguardian** [comment]
Jenkins retest this please.

File: kie-server-parent/kie-server-remote/kie-server-client/src/build/revapi-config.json [hunk 1]
```diff
@@ -41,7 +41,63 @@
...
+                },
+                {
+                    "code": "java.method.addedToInterface",
+                    "new": "method java.util.List<org.kie.server.api.model.cases.CaseInstance> org.kie.server.client.CaseServicesClient::getCaseInstancesByContainer(java.lang.String, java.util.List<java.lang.String>, java.lang.Integer, java.lang.Integer, java.lang.String, boolean, boolean)",
+                    "package": "org.kie.server.client",
+                    "classSimpleName": "CaseServicesClient",
+                    "methodName": "getCaseInstancesByContainer",
+                    "elementKind": "method",
+                    "justification": "https://issues.redhat.com/browse/JBPM-9223"
+                },
```
Comment (afalhambra) [inline kie-server-parent/kie-server-remote/kie-server-client/src/build/revapi-config.json:80]: These is not needed. These changes are already in 7.44 final version.
```suggestion
```

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/build/revapi-config.json [hunk 2]
```diff
@@ -28,7 +28,278 @@
...
+                   "old": "method javax.ws.rs.core.Response org.kie.server.remote.rest.jbpm.RuntimeDataResource::queryProcessesByVariables(javax.ws.rs.core.HttpHeaders, java.lang.String, java.lang.Integer, java.lang.Integer)",
+                   "new": "method javax.ws.rs.core.Response org.kie.server.remote.rest.jbpm.RuntimeDataResource::queryProcessesByVariables(javax.ws.rs.core.HttpHeaders, java.lang.String, java.lang.Integer, java.lang.Integer)",
+                   "annotationType": "io.swagger.annotations.ApiOperation",
+                   "annotation": "@io.swagger.annotations.ApiOperation(\"Queries processes by variables and tasks\")",
+                   "package": "org.kie.server.remote.rest.jbpm",
+                   "classSimpleName": "RuntimeDataResource",
+                   "methodName": "queryProcessesByVariables",
+                   "elementKind": "method",
+                   "justification": "JBPM-9296 Self contained Audit/History module"
+                 },
```
Comment (afalhambra) [inline kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/build/revapi-config.json:142]: This ignore section is not needed, changes are already in 7.44. final version. Please remove.
```suggestion
```

**Review by afalhambra — CHANGES_REQUESTED** [review]
Very nice work, just a few comments and request. Thanks

File: kie-server-parent/kie-server-remote/kie-server-client/src/build/revapi-config.json (hunk 1 above)
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-client/src/build/revapi-config.json:80]: done

File: kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/impl/ProcessServicesClientImpl.java [hunk 3]
```diff
@@ -592,6 +594,56 @@ public void signalProcessInstances(String containerId, List<Long> processInstanc
...
+            throwExceptionOnFailure(response);
+        }
+    }
+
+    @Override
+    public void signalProcessInstancesByCorrelationKeys(String containerId,
+                                                        List<CorrelationKey> correlationKeys,
+                                                        String signalName,
+                                                        Object event) {
+        List<String> keys = correlationKeys.stream().map(e -> e.toExternalForm()).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/impl/ProcessServicesClientImpl.java:625]: done

File: kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/impl/ProcessServicesClientImpl.java [hunk 4]
```diff
@@ -592,6 +594,56 @@ public void signalProcessInstances(String containerId, List<Long> processInstanc
...
+            String queryStr = buildQueryString("correlationKeys", keys);
+
+
+            Map<String, String> headers = new HashMap<String, String>();
+            makeHttpPostRequestAndCreateCustomResponse(
+                    build(loadBalancer.getUrl(), PROCESS_URI + "/" + SIGNAL_PROCESS_INSTANCES_PORT_URI, valuesMap) + queryStr
+                    , event, String.class, headers);
+        } else {
+            CommandScript script = new CommandScript(singletonList(
+                    (KieServerCommand) new DescriptorCommand( "ProcessService", "signalProcessInstancesByCorrelationKey", serialize(event), marshaller.getFormat().getType(), new Object[]{containerId, keys, signalName})));
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/impl/ProcessServicesClientImpl.java:641]: This is correct. This has to do with the naming in the process service base.

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/ProcessResource.java [hunk 5]
```diff
@@ -430,7 +431,44 @@ public Response signalProcessInstance(@javax.ws.rs.core.Context HttpHeaders head
...
+            @ApiResponse(code = 404, message = "Process instance or Container Id not found"),
+            @ApiResponse(code = 403, message = "User does not have permission to access this asset")})
+    @POST
+    @Path(SIGNAL_PROCESS_BY_CORRELATION_KEY_POST_URI)
+    @Consumes({MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML})
+    @Produces({MediaType.APPLICATION_JSON, MediaType.APPLICATION_XML})
+    public Response signalProcessInstanceByCorrelationKey(@javax.ws.rs.core.Context HttpHeaders headers, 
+            @ApiParam(value = "container id that process instance belongs to", required = true, example = "evaluation_1.0.0-SNAPSHOT") @PathParam(CONTAINER_ID) String containerId,
+            @ApiParam(value = "correlation key of the process instance to be signaled", required = true, example = "123") @PathParam(CORRELATION_KEY) String correlationKey, 
+            @ApiParam(value = "signal name to be send to process instance", required = true, example = "EventReceived") @PathParam(SIGNAL_NAME) String signalName,
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/main/java/org/kie/server/remote/rest/jbpm/ProcessResource.java:446]: done

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/ProcessServiceBase.java [hunk 6]
```diff
@@ -230,6 +231,29 @@ public void signalProcessInstances(String containerId, List<Long> processInstanc
...
+                                                      String correlationKey,
+                                                      String signalName,
+                                                      String eventPayload,
+                                                      String type) {
+
+        String[] correlationProperties = correlationKey.split(":");
+        CorrelationKey actualCorrelationKey = correlationKeyFactory.newCorrelationKey(Arrays.asList(correlationProperties));
+
+        logger.debug("Calling signal '{}' process instances with correlation key {} on container {} and event {}", signalName, correlationKey, containerId, null);
+        processService.signalProcessInstanceByCorrelationKey(actualCorrelationKey, signalName, null);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/ProcessServiceBase.java:244]: This is the parameter convention used by JMS protocol so it is not possible to remove it. It has to be this way. (it is used by reflection)

File: kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/ProcessServiceBase.java [hunk 7]
```diff
@@ -230,6 +231,29 @@ public void signalProcessInstances(String containerId, List<Long> processInstanc
...
+    }
+
+    public void signalProcessInstancesByCorrelationKey(String containerId,
+                                                       List<String> correlationKeys,
+                                                       String signalName,
+                                                       String eventPayload,
+                                                       String type) {
+        List<CorrelationKey> keys = correlationKeys.stream().map(e -> correlationKeyFactory.newCorrelationKey(Arrays.asList(e.split(":")))).collect(Collectors.toList());
+        logger.debug("Calling signal '{}' process instances with correlation key {} on container {} and event {}", signalName, keys, containerId, null);
+        processService.signalProcessInstancesByCorrelationKeys(keys, signalName, null);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-services/kie-server-services-jbpm/src/main/java/org/kie/server/services/jbpm/ProcessServiceBase.java:254]: This is JMS client protocol (same as the other.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java [hunk 8]
```diff
@@ -273,6 +273,67 @@ public void testSignalProcessInstance() throws Exception {
         }
     }
     
+    @Test
+    public void testSignalProcessInstanceByCorrelationKey() throws Exception {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java:277]: removed

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java [hunk 9]
```diff
@@ -273,6 +273,67 @@ public void testSignalProcessInstance() throws Exception {
...
+        Long processInstanceId = processClient.startProcess(CONTAINER_ID, PROCESS_ID_SIGNAL_PROCESS);
+
+        assertNotNull(processInstanceId);
+        assertTrue(processInstanceId.longValue() > 0);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java:281]: done

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java [hunk 10]
```diff
@@ -273,6 +273,67 @@ public void testSignalProcessInstance() throws Exception {
...
+
+            processClient.signalProcessInstanceByCorrelationKey(CONTAINER_ID, correlationKey, "Signal2", "My custom string event");
+        } catch (Exception e) {
+            processClient.abortProcessInstance(CONTAINER_ID, processInstanceId);
+            fail(e.getMessage());
+        }
+    }
+
+    @Test
+    public void testSignalProcessInstancesByCorrelationKeys() throws Exception {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java:300]: removed

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java [hunk 11]
```diff
@@ -273,6 +273,67 @@ public void testSignalProcessInstance() throws Exception {
...
+        Long processInstanceId = processClient.startProcess(CONTAINER_ID, PROCESS_ID_SIGNAL_PROCESS);
+
+        assertNotNull(processInstanceId);
+        assertTrue(processInstanceId.longValue() > 0);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java:304]: removed

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java [hunk 12]
```diff
@@ -273,6 +273,67 @@ public void testSignalProcessInstance() throws Exception {
...
+
+            Object person = createPersonInstance(USER_JOHN);
+            processClient.signalProcessInstanceByCorrelationKey(CONTAINER_ID, correlationKey, "Signal1", person);
+
+            processClient.signalProcessInstanceByCorrelationKey(CONTAINER_ID, correlationKey, "Signal2", "My custom string event");
+        } catch (Exception e) {
+            processClient.abortProcessInstance(CONTAINER_ID, processInstanceId);
+            fail(e.getMessage());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java:297]: assert that process is finished. only can happen when both signal were processed.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java [hunk 13]
```diff
@@ -273,6 +273,67 @@ public void testSignalProcessInstance() throws Exception {
...
+    public void testSignalProcessInstancesByCorrelationKeys() throws Exception {
+        Long processInstanceId = processClient.startProcess(CONTAINER_ID, PROCESS_ID_SIGNAL_PROCESS);
+
+        assertNotNull(processInstanceId);
+        assertTrue(processInstanceId.longValue() > 0);
+        ProcessInstance pi1 = processClient.getProcessInstance(CONTAINER_ID, processInstanceId);
+
+        Long processInstanceId2 = processClient.startProcess(CONTAINER_ID, PROCESS_ID_SIGNAL_PROCESS);
+        assertNotNull(processInstanceId2);
+        assertTrue(processInstanceId2.longValue() > 0);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java:309]: assert that process is finished. only can happen when both signal were processed.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java [hunk 14]
```diff
@@ -273,6 +273,67 @@ public void testSignalProcessInstance() throws Exception {
...
+
+            Object person = createPersonInstance(USER_JOHN);
+            processClient.signalProcessInstancesByCorrelationKeys(CONTAINER_ID, correlationKeys, "Signal1", person);
+
+            processClient.signalProcessInstancesByCorrelationKeys(CONTAINER_ID, correlationKeys, "Signal2", "My custom string event");
+        } catch (Exception e) {
+            processClient.abortProcessInstance(CONTAINER_ID, processInstanceId);
+            processClient.abortProcessInstance(CONTAINER_ID, processInstanceId2);
+            fail(e.getMessage());
+        }
```
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java:333]: assert that process is finished. only can happen when both signal were processed.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (elguardian) [reply kie-server-parent/kie-server-tests/kie-server-integ-tests-jbpm/src/test/java/org/kie/server/integrationtests/jbpm/ProcessServiceIntegrationTest.java:309]: done

File: kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/build/revapi-config.json (hunk 2 above)
Comment (elguardian) [reply kie-server-parent/kie-server-remote/kie-server-rest/kie-server-rest-jbpm/src/build/revapi-config.json]: removed.

**Review by afalhambra — COMMENTED** [review]
Not sure about this, but it seems there are changes in new files which I don't think they are related to this PR/BAPL.

These are:

- https://github.com/kiegroup/droolsjbpm-integration/blob/6c5f5e68fcdf5d6770c80a63767902444d67e28d/kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/Configuration.java
- https://github.com/kiegroup/droolsjbpm-integration/blob/6c5f5e68fcdf5d6770c80a63767902444d67e28d/kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/ContainerInfo.java
- https://github.com/kiegroup/droolsjbpm-integration/blob/6c5f5e68fcdf5d6770c80a63767902444d67e28d/kie-server-parent/kie-server-router/kie-server-router-proxy/src/main/java/org/kie/server/router/repository/ConfigurationMarshaller.java
- https://github.com/kiegroup/droolsjbpm-integration/blob/6c5f5e68fcdf5d6770c80a63767902444d67e28d/kie-server-parent/kie-server-router/kie-server-router-proxy/src/test/java/org/kie/server/router/ConfigurationTest.java

Am I missing something here?

**Comment by elguardian** [comment]
@afalhambra removed. mistake with the command line

**Review by afalhambra — APPROVED** [review]
Good work @elguardian - thanks

**Pull request merged**